tee-worker/service/src/main_impl.rs

```diff
@@ -505,110 +505,103 @@ fn start_worker<E, T, D, InitializationHandler, WorkerModeProvider>(
 	#[cfg(feature = "dcap")]
 	enclave.set_sgx_qpl_logging().expect("QPL logging setup failed");
 
-	// Litentry: send the registration extrinsic regardless of being registered or not,
-	//           the reason is the mrenclave could change in between, so we rely on the
-	//           on-chain logic to handle everything.
-	//           this is the same behavior as upstream
-	let register_enclave_block_hash = register_enclave(
-		enclave.clone(),
-		&litentry_rpc_api,
-		&tee_accountid,
-		&trusted_url,
-		skip_ra,
-		is_development_mode,
-	)
-	.expect("enclave RA registration must be successful to continue");
+	// Litentry:
+	// the logic differs from upstream a bit here (due to different impl in parachain pallet),
+	// `primary_enclave_identifier_for_shard` returns None if there is no worker registered for the shard.
+	//
+	// The logic here is as follows:
+	// in the case it's Some, it means there is a primary worker registered for the shard, we are a non-primary worker if it is first run
+	// so we need to request provisioning. Finally in any case we register worker.
+	//
+	// in case it's None, it means there is no primary worker registered for the shard, then we are the primary worker
+	//
+	// based on the check of `enclave.get_shard_creation_info` to tell if this worker has run before - this is similar to upstream.
+	// There're a few cases:
+	// 1. `--clean-reset` is set, then the shard should have been initalized earlier already and it's empty state anyway
+	// 2. `--clean-reset` is not set:
+	//    2a. `get_shard_creation_info` is empty and we are primary worker => it's never run before => init everything
+	//    2b. `get_shard_creation_info` is empty and we are non-primary worker => it's never run before => request to sync state
+	//    2c. `get_shard_creation_info` is non-empty it's run before => do nothing
+	let primary_enclave_id = litentry_rpc_api
+		.primary_enclave_identifier_for_shard(WorkerType::Identity, shard, None)
+		.expect("primary_enclave_identifier_for_shard failed");
+
+	let mut register_enclave_xt_header: Header;
+
+	let (we_are_primary_validateer, re_init_parentchain_needed) = match primary_enclave_id {
+		Some(primary_account_id) => {
+			println!("Primary enclave for shard is: {:?}", primary_account_id);
+			let is_first_run = enclave
+				.get_shard_creation_info(shard)
+				.unwrap()
+				.for_parentchain(ParentchainId::Litentry)
+				.is_none();
+
+			if is_first_run {
+				// we are not primary worker
```

Comment:
So trying to understand the logic here, We are not checking if `primary_account_id` is equal to `tee_account_id` or not, and we are assuming `is_first_run` also means that it is a non-primary worker? If `tee_account_id` is equal to `primary_account_id` are we effectively trying to perform `sync state` with ourselves? Let me know if i'm missing something

Reply:
In the previous logic, we first registered the enclave, then we checked if there was some enclave registered (the first one, so the primary) and if it was the first run or not to run the init shard logic. With these changes we first try to fetch the primary enclave (before registering), then:

- if `primary_enclave_id` is `Some`, it can be two things:
1. It's the first run, so we assume we are not the primary worker because there is already an enclave registered.  (We could also double check that `tee_accountid` is different than the `primary_account_id`, but I guess it's redundant at this point?). So we sync state.
2. It's not first run, so it may be the primary worker. (In this case there is no additional logic we need to run).
  
   Then we register the enclave.
  
- if `primary_enclave_id` is `None`, there is no enclave registered so we are the primary worker. We register the enclave and run the init shard logic. 

Does this makes sense?

Reply:
I updated the code to double check the cases where we are not the primary worker to sync state

Reply:
Ah, I see, Totally makes sense.